### ISSUE TITLE
"Title" attribute as a valid alternative for "Hidden Label"

### ISSUE THREAD

**Issue by mdkerner** [UCSantaBarbara/webguide#30]
I was updating the Admissions site with some accessibility fixes when I ran across this recommendation from AMP:

Provide valid labels for form fields
Description
When on-screen labels are present, they must be explicitly associated with form fields. When on-screen labels are not present, form fields must be given an accessible label.
When form fields do not explicitly include a label, assistive technologies may incorrectly render the label or provide no label at all to users. When labels are not present or are incorrect, users of assistive technologies may not be able to complete a form.
The HTML5 specification has a new form field attribute called placeholder. This represents a label or hint, such as a word or short phrase, that is assigned to a form field such as an input field. The label or hint appears within the form field and goes away when users start typing. When the placeholder attribute is used, the label or hint may not be detected by assistive technologies. Therefore, developers should provide off-screen text in the label element of the form field **or provide a title**

I looked further into this, and found this on W3C:

http://www.w3.org/TR/WCAG20-TECHS/H65.html

The objective of this technique is to use the title attribute to label form controls when the visual design cannot accommodate the label (for example, if there is no text on the screen that can be identified as a label) or where it might be confusing to display a label. User agents, including assistive technology, can speak the title attribute.

Thoughts? Concerns? Rebuttals?

**Comment by ericmayes** [comment]
My recommendation would be to use a label whenever possible. #1, placeholders should never be used instead of labels but can be used in addition to a label for the reason described above. I try to avoid using the title attribute since tooltips are annoying to users, plus if a label and a title are both used the screen reader will read both. If the label must be hidden (which isn't good for sighted users or those with cognitive disabilities) then a hidden label is my recommendation. But sometimes elements hidden off screen can cause issues on mobile screen readers but returning to the top of the page after reading the off screen element. So basically, a normal label should be used if possible. It makes more sense to everyone. Alternatively you could also use an aria-label, how ever the AMP tool may not get rid of the error with the use of the aria-label (hence why the UC is looking for a newer tool). http://www.pauljadam.com/demos/title-aria-label.html  check the examples on this page and you can see the results for different screen reader and browser combinations.

**Comment by mdkerner** [comment]
I just dislike the "hackiness" of having an element on the page, then pushing it waaaaaay off the screen. It's just something that's never sat right with me. I saw an alternative and got really excited about it.

Also, I'd never heard of aria-labels before. This sounds like a solution that I'd be okay with (I still hate hidden labels.)

**Comment by ericmayes** [comment]
I totally agree. Title should work fine if you look at the results from the link I sent. I don't really like the tooltip a title generates. But it is a valid solution as well. Paul Adams results with modern mobile screen readers show that it works now so it should be ok. The title tag will probably make your automated test pass and also work in all AT, not sure if the aria-label will change the score of the automated test. However I second guess the title attribute again because of articles like this (a bit outdated though) that talk about user groups that the title attrib doesnt serve well: http://www.paciellogroup.com/blog/2010/11/using-the-html-title-attribute/

also here is another good article on use and abuse of the title attrib:
http://www.paciellogroup.com/blog/2012/01/html5-accessibility-chops-title-attribute-use-and-abuse/

**Comment by loganfranken** [comment]
What an awesome discussion!

I've seen this need to support label-less inputs come up in two use cases:

### 1: Aesthetics

The label has been removed because it looks nicer without it. For example, the CSS Tricks comment form:

![comment-field](https://cloud.githubusercontent.com/assets/1024444/7886235/8e1035c6-05dd-11e5-9fa1-1793ac249ab2.png)

### 2: Redundant

The label is redundant because the form makes it obvious, like the current WSG search:

``` html
<form class="search-form" method="GET" action="/search">
   <label for="search-query" class="hidden">Search Query:</label>
   <input type="text" id="search-query" name="query" />
   <input class="search-btn" type="submit" value="Search" />
</form>
```

In this case, it's obvious that you're supposed to enter a search query given the `value` of the submit button is "Search" so the label becomes redundant.

Rich Trott (@trott) from UCSF also explored this [issue](https://twitter.com/trott/status/494288796968841218).

The answer he received on [StackOveflow](http://ux.stackexchange.com/questions/62059/omiting-label-element-on-self-explanatory-form/62147) surprised me:

(According to [WCAG 2.0 G167](http://www.w3.org/TR/2014/NOTE-WCAG20-TECHS-20140408/G167))

"When a button invokes a function on an input field, has a clear text label, and is rendered adjacent to the input field, the button also acts as a label for the input field. This label helps users understand the purpose of the field without introducing repetitive text on the Web page."

### Testing

I would actually be interested in doing some tests to see how the various options actually perform under the screenreader microscope.

Anyone got any other use cases they would like to test?

**Comment by loganfranken** [comment]
I put together a test page to try this out for myself:
http://loganfranken.github.io/web-standards-testing/title-as-label/

**Comment by loganfranken** [comment]
Alright, I've done some initial testing with NVDA on Windows 8.1:
http://loganfranken.github.io/web-standards-testing/title-as-label/results.html

Results were pretty surprising:

An actual `label`, a `label` hidden with CSS, a `title` attribute, and `aria-label` all performed just as well when I focused into the element: in all three cases, the value stored in that attribute was read aloud by NVDA in latest Chrome/FF/IE.

The one caveat (as @ericmayes mentions above) is that the `title` attribute causes a tooltip to appear.

I'll try testing with JAWS and the others too.

If results are the same with other screen readers, I'm leaning towards recommending `aria-label` if a `label` is not available since `title` has the tooltip side effect mentioned above.

**Comment by ericmayes** [comment]
Great work Logan! I ran your test page through iOS 8.3 Voiceover with the following results:

input and label
Results
iOS Voiceover   "Name, Text Field, Double tap to edit"

label hidden with CSS
Results
iOS Voiceover   "Name, Text Field, Double tap to edit"

No label
Results
iOS Voiceover   "Text Field, Double tap to edit"

No label with title attribute
Results
iOS Voiceover   "Name, Text Field, Double tap to edit"

No label with placeholder attribute
Results
iOS Voiceover   "Enter Name, Text Field, Double tap to edit"

No label with aria-label attribute
Results
iOS Voiceover   "Name, Text Field, Double tap to edit"

Implied label
Results
iOS Voiceover   "Text Field, Double tap to edit"

**Comment by loganfranken** [comment]
You're awesome, @ericmayes! Thank you so much!

Wow, I totally did not expect `title` to hold up this well.

**Comment by ericmayes** [comment]
Me neither! I think its because the latest browsers and AT have better support for it now.

**Comment by loganfranken** [comment]
Wow, so I just tried this in JAWS 16.0.3048 and got the same exact results as NVDA.

**Comment by loganfranken** [comment]
Hey @ericmayes, when you get a second: were you testing in Safari? Which version?

(Just want to update the results page: http://loganfranken.github.io/web-standards-testing/title-as-label/results.html)

**Comment by ericmayes** [comment]
Yes Safari for iOS 8.3 Just for kicks I tested in Safari 8.05 in OSX 10.10.3 as well (OSX Voiceover)
input and label
Results
Safari 8.05 "Name, Edit Text"

label hidden with CSS
Results
Safari 8.05 "Name, Edit Text"

No label
Results
Safari 8.05 "Edit Text, blank"

No label with title attribute
Results
Safari 8.05 "Name, Edit Text"

No label with placeholder attribute
Results
Safari 8.05 "Enter Name, Edit Text"

No label with aria-label attribute
Results
Safari 8.05 "Name, Edit Text"

Implied label
Results
Safari 8.05 "Edit Text, blank"

**Comment by loganfranken** [comment]
Awesome! Thanks @ericmayes!

Updated the results page with that additional platform:
http://loganfranken.github.io/web-standards-testing/title-as-label/results.html

Whatever recommendation we decide on, I think it would look a little out-of-place if we plopped this on the [accessibility page](http://webguide.ucsb.edu/web-accessibility/) since it's mostly just an introduction to the relevant policies with several links to PDFs.

@tenken How do you feel about creating a wiki for this repo? Maybe we could start putting together a list of accessibility tips that could be the basis for an eventual update of the [accessibility guidelines PDF](http://webguide.ucsb.edu/media/web-accessibility-guidelines.pdf)?

**Comment by tenken** [comment]
I have created a new wiki with the [Accessibility Tips](../wiki/Accessibility-Tips) subpage.

FYI, i recommend linking to wiki pages from issues as:

<pre>
 [Accessibility Tips](../wiki/Accessibility-Tips)
</pre>

This will survive repository renames and custom domain changes as it's relative.

Anyone if free to contribute to the Wiki, but I would love to preserve the existing workflow of the Webguide and hope that contributions to the Wiki (content) and Webguide (content) remain appropriate.

**Comment by loganfranken** [comment]
@tenken Are we going to edit the wiki directly? Or did you want it to happen via pull request?

(I can't edit myself currently)

**Comment by tenken** [comment]
I think I just opened up the wiki to be editable by the public. Assuming it's not flooded with inappropriate content lets just try this configuration for the Wiki for the time being.

Please test and supply feedback, thanks.

**Comment by loganfranken** [comment]
I've updated the wiki with my summary of findings on this topic:
https://github.com/ucsb-wsg/ucsb-wsg.github.io/wiki/Accessibility-Tips#always-use-a-label-or-accessible-alternative-for-input-fields

(Essentially: always use a `label`. If you can't, use `aria-label`)

In other words, I _won't_ suggest using `title` (even though it's proven a worthy alternative) since it has some undesirable side effects (tooltips) and `aria-label` is more specifically built for this use case.

I'm going to keep researching this topic a little further and I'll post any updates here. I think we can let this topic linger a little longer for additional opinions and then close it.

**Comment by tenken** [comment]
@loganfranken Thanks.  (I need to start using these @mention better) :P

**Comment by mdkerner** [comment]
@loganfranken 

I was waiting until there was a final decision regarding 'title vs aria-label'! I'm taking this as "Recommended Best Practice" now and I will hear no more about it unless I am forced to. Have a good day gentlepeople!
